contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,847 @@
+    // controller from the PodTemplateSpec indicated by CurrentTemplateGeneration.
+ 	CurrentReplicas int32 `json:"currentReplicas,omitempty"`
+ 	
+ 	// UpdatedReplicas is the number of Pods created by the StatefulSet
+    // controller from the PodTemplateSpec indicated by TargetTemplateGeneration.
+ 	UpdatedReplicas int32 `json:"taretReplicas,omitempty"`
+}
+```
+
+Additionally, we introduce the following constant.
+
+```go 
+// StatefulSetTemplateGenerationLabel is the label applied to a PodTemplate or 
+// Pod to indicate the position of the object's Template in the revision 
+// history of a StatefulSet.
+const StatefulSetTemplateGenerationLabel = "statefulset-template-revision"
+```
+
+## StatefulSet Controller
+The StatefulSet controller will watch for modifications to StatefulSet and Pod 
+API objects. When a StatefulSet is created or updated, or when one 
+of the Pods in a StatefulSet is updated or deleted, the StatefulSet
+controller will attempt to create, update, or delete Pods to conform the 
+current state of the system to the user declared target state. 
+The user declared target state of the system, with respect to an individual 
+StatefulSet, is determined as below.
+
+### Target State
+The declared target state of a StatefulSet requires that all Pods in the 
+StatefulSet conform to exactly one or two PodTemplates in the StatefulSet's 
+revision history. If the declared target state references two PodTemplates, as 
+is the case when a user wants to perform a canary update or a phased roll out, 
+they are partitioned around an ordinal such that all Pods with a lower ordinal 
+conform to one PodTemplate and all Pods with a greater or equal ordinal 
+conform to the other. The conditions that define this state in terms of the 
+StatefulSet's StatefulSetSpec and StatefulSetStatus are below.
+
+1. The StatefulSet contains exactly `[0,.Spec.Replicas)` Pods.
+1. If StatefulSet's `.Spec.GenerationPartition` is nil, then the following is 
+true.
+    1. The StatefulSet's `.Status.CurrentTemplateGeneration` is equal to its 
+    `.Status.UpdatedTemplateGeneration`.
+    1. All Pods in the StatefulSet have been generated from the PodTemplate 
+    labeled with a `StatefulSetTemplateGenerationLabel` equal to its 
+    `.Status.TemplateGeneration`.
+1. If the StatefulSet's `.Spec.GenerationPartition` is not nil, then the 
+following is true.
+    1. All Pods with ordinals is the sequence `[0,.Spec.GenerationPartition)` have 
+     been generated from the PodTemplate in the StatefulSet's revision history
+     that is labeled with a `StatefulSetTemplateGenerationLabel` equal to 
+     `.Status.TemplateGeneration`.
+    1. All Pods with ordinals in the sequence 
+     `[Spec.GenerationParition,.Spec.Replicas)` have been created with the 
+     PodTemplate in the StatefulSet's revision history that is labeled with a 
+     `StatefulSetTemplateGenerationLabel` equal to 
+     `.Status.UpdatedTemplateGeneration`.
```

Comment:
@kow3ns i am still confused where you keep the different revisions of the pod template ?